DOC: Convert to markdown: torch.compiler_best_practices_for_backends.rst, torch.compiler_cudagraph_trees.rst, torch.compiler_custom_backends.rst, torch.compiler_dynamic_shapes.rst, torch.compiler_dynamo_deepdive.rst

aagalleg opened the pull request:
Fixes #155037 

[torch.compiler_best_practices_for_backends.rst](https://github.com/pytorch/pytorch/tree/main/docs/source/torch.compiler_best_practices_for_backends.rst) shows error 404

cc  @svekars @sekyondaMeta @AlannaBurke

aagalleg committed:
Change extension from rst to md

- Change extension from: torch.compiler_cudagraph_trees.rst, torch.compiler_custom_backends.rst, torch.compiler_dynamic_shapes.rst, torch.compiler_dynamo_deepdive.rst

aagalleg committed:
Convert from rst to md formatting in CUDAGraph Trees documentation

- Updated headings, code blocks, links, and tables to Markdown syntax.

aagalleg committed:
Convert from rst to md formatting in Custom Backends documentation

- Updated headings, code blocks and links to Markdown syntax.

aagalleg committed:
Convert from rst to md formatting in Dynamic Shapes documentation

- Updated headings, code blocks and links to Markdown syntax.

aagalleg committed:
Convert from rst to md formatting in Dynamo Deep-Dive documentation

- Updated headings, code blocks and links to Markdown syntax.

aagalleg commented:
@pytorchbot label "topic: not user facing"

svekars commented:
In the diff, it shows that docs/source/torch.compiler_custom_backends.rst was deleted and  docs/source/torch.compiler_custom_backends.md was created. Can you please make sure you use git mv for converting these files.

aagalleg commented:
> In the diff, it shows that docs/source/torch.compiler_custom_backends.rst was deleted and docs/source/torch.compiler_custom_backends.md was created. Can you please make sure you use git mv for converting these files.

Hi.
Yes, I'm sure that I used the command. Commit a77a3b1e205a570d4bf87fabb26ae6df562b94ba says that files were "File renamed without changes".

aagalleg committed:
Add removed anchors

- Adding anchors torch.compiler_dynamo_deepdive and making-dynamo-sound-guards that were removed by mistake.

zxiiro commented:
> > In the diff, it shows that docs/source/torch.compiler_custom_backends.rst was deleted and docs/source/torch.compiler_custom_backends.md was created. Can you please make sure you use git mv for converting these files.
> 
> Hi. Yes, I'm sure that I used the command. Commit [a77a3b1](https://github.com/pytorch/pytorch/commit/a77a3b1e205a570d4bf87fabb26ae6df562b94ba) says "File renamed without changes".

Yeah. Git diff tracks contents of files not filenames so when 50% or more of the lines in a file is changed Git diff will report it as a deleted and new file rather than a rename. On the commandline you can use `git diff -M50%` to adjust the threshold.

svekars commented:
ok, it looks like the history is being tracked. Can you fix the doc build error?

aagalleg commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -1,9 +1,7 @@
-.. _torch.compiler_dynamo_deepdive:
+(torch.compiler_dynamo_deepdive)=
+# Dynamo Deep-Dive
```
Done.

aagalleg committed:
Move anchor in Dynamo Deep-Dive

- Change position of anchor to fix doc build.

aagalleg commented:
> ok, it looks like the history is being tracked. Can you fix the doc build error?

It looks like moving the anchor doesn't solve the problem. I tried correcting the error by moving the anchor to other sections but it keeps not being referenced in my local server.

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -1,9 +1,7 @@
-.. _torch.compiler_dynamo_deepdive:
+# Dynamo Deep-Dive
+(torch.compiler_dynamo_deepdive)=
```
try to make it a paragraph target?
```suggestion

(#torch.compiler_dynamo_deepdive)=
```

svekars committed:
Update docs/source/torch.compiler_dynamo_deepdive.md

svekars committed:
Update docs/source/torch.compiler_dynamo_deepdive.md

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -1,9 +1,10 @@
 
-Dynamo Deep-Dive
-================
+(#torch.compiler_dynamo_deepdive)=
```
```suggestion
```

svekars committed:
Update docs/source/torch.compiler_dynamo_deepdive.md

aagalleg committed:
Fix ERROR: Document or section may not begin with a transition.

- Footnotes section marked an error when inmediate text is a footnote that starts with [^footnote]
- Added line prevents error from appearing.
- After adding the new line, (torch.compiler-dynamo-deepdive)= anchor placement didn't cause problems.

aagalleg committed:
Add blockquote to Note in CUDAGraph Trees

- To match the style of the original rst file a blockquote is added.

svekars committed:
Add a blank line between the title and the target

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -213,155 +210,151 @@ to execute the function. In order to be able to execute the function, on
 landed at
-`_convert_frame_assert <https://github.com/pytorch/pytorch/blob/b6df8414601e1e086e830ca9e919e7fdc8874e71/torch/_dynamo/convert_frame.py#L272-L274>`__.
-This is the function that the decorator ``torch.compile`` returns! We
+[\_convert_frame_assert](https://github.com/pytorch/pytorch/blob/b6df8414601e1e086e830ca9e919e7fdc8874e71/torch/_dynamo/convert_frame.py#L272-L274).
```
```suggestion
[_convert_frame_assert](https://github.com/pytorch/pytorch/blob/b6df8414601e1e086e830ca9e919e7fdc8874e71/torch/_dynamo/convert_frame.py#L272-L274).
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -213,155 +210,151 @@ to execute the function. In order to be able to execute the function, on
-`_dynamo.optimize <https://github.com/pytorch/pytorch/blob/b6df8414601e1e086e830ca9e919e7fdc8874e71/torch/_dynamo/eval_frame.py#L715-L727>`__.
-The decorator ``torch.compile`` is just a nice API around
-``_dynamo.optimize``.
+[\_dynamo.optimize](https://github.com/pytorch/pytorch/blob/b6df8414601e1e086e830ca9e919e7fdc8874e71/torch/_dynamo/eval_frame.py#L715-L727).
```
```suggestion
[_dynamo.optimize](https://github.com/pytorch/pytorch/blob/b6df8414601e1e086e830ca9e919e7fdc8874e71/torch/_dynamo/eval_frame.py#L715-L727).
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -213,155 +210,151 @@ to execute the function. In order to be able to execute the function, on
+[TensorVariable](https://github.com/pytorch/pytorch/blob/83c0763dda1f93c6cf552ba88260a0dc7a3ecb70/torch/_dynamo/variables/tensor.py#L68-L69).
 All these internal classes are defined in the
-`torch/_dynamo/variables <https://github.com/pytorch/pytorch/tree/83c0763dda1f93c6cf552ba88260a0dc7a3ecb70/torch/_dynamo/variables>`__
+[torch/\_dynamo/variables](https://github.com/pytorch/pytorch/tree/83c0763dda1f93c6cf552ba88260a0dc7a3ecb70/torch/_dynamo/variables)
```
```suggestion
[torch/_dynamo/variables](https://github.com/pytorch/pytorch/tree/83c0763dda1f93c6cf552ba88260a0dc7a3ecb70/torch/_dynamo/variables)
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -213,155 +210,151 @@ to execute the function. In order to be able to execute the function, on
 class in
-`VariableBuilder._wrap <https://github.com/pytorch/pytorch/blob/83c0763dda1f93c6cf552ba88260a0dc7a3ecb70/torch/_dynamo/variables/builder.py#L365>`__.
-This function is just a very long chain of ``elif``\ s that tries to
+[VariableBuilder.\_wrap](https://github.com/pytorch/pytorch/blob/83c0763dda1f93c6cf552ba88260a0dc7a3ecb70/torch/_dynamo/variables/builder.py#L365).
```
```suggestion
[VariableBuilder._wrap](https://github.com/pytorch/pytorch/blob/83c0763dda1f93c6cf552ba88260a0dc7a3ecb70/torch/_dynamo/variables/builder.py#L365).
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -435,48 +428,48 @@ local or global variables present when entering the current frame. In
 All these are defined in
-`torch/_dynamo/source.py <https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/source.py>`__.
-We can see the guard generated by ``GetItemSource`` in the following
+[torch/\_dynamo/source.py](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/source.py).
```
```suggestion
[torch/_dynamo/source.py](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/source.py).
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -488,24 +481,22 @@ why we need to have such fine control over the Python interpreter as to
-it in the output FX graph. [4]_ We already saw this class in the introduction
+[torch.SymInt](https://github.com/pytorch/pytorch/blob/fb80f05ee2e1cba17892980701bfd5dbce58349f/torch/__init__.py#L244-L249)
+that acts like an `int` but it records all the operations performed on
+it in the output FX graph. [^footnote-4] We already saw this class in the introduction
```
```suggestion
it in the output FX graph. [^4] We already saw this class in the introduction
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -213,155 +210,151 @@ to execute the function. In order to be able to execute the function, on
 
 You can see all the fields
-`here <https://github.com/pytorch/pytorch/blob/e891a3bba9f05697d72776f6e89347231a141f03/torch/csrc/dynamo/eval_frame.c#L50-L59>`__. [2]_
+[here](https://github.com/pytorch/pytorch/blob/e891a3bba9f05697d72776f6e89347231a141f03/torch/csrc/dynamo/eval_frame.c#L50-L59). [^footnote-2]
```
```suggestion
[here](https://github.com/pytorch/pytorch/blob/e891a3bba9f05697d72776f6e89347231a141f03/torch/csrc/dynamo/eval_frame.c#L50-L59). [^2]
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -213,155 +210,151 @@ to execute the function. In order to be able to execute the function, on
 
 In summary, CPython provides the user’s interpreter with all the
-information necessary to execute the function. [3]_
+information necessary to execute the function. [^footnote-3]
```
```suggestion
information necessary to execute the function. [^3]
```

AlannaBurke commented:
https://docs-preview.pytorch.org/pytorch/pytorch/155137/torch.compiler_dynamic_shapes.html looks good.

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -213,155 +210,151 @@ to execute the function. In order to be able to execute the function, on
-`InstructionTranslator object <https://github.com/pytorch/pytorch/blob/69f112d5867f785a3a090a0c6d6644ae047033ac/torch/_dynamo/symbolic_convert.py#L2060-L2071>`__
+[OutputGraph](https://github.com/pytorch/pytorch/blob/69f112d5867f785a3a090a0c6d6644ae047033ac/torch/_dynamo/output_graph.py#L221-L230)
+object. The `OutputGraph` object is [bound to an
+\`InstructionTranslator object](https://github.com/pytorch/pytorch/blob/69f112d5867f785a3a090a0c6d6644ae047033ac/torch/_dynamo/symbolic_convert.py#L2060-L2071)
```
```suggestion
`InstructionTranslator object](https://github.com/pytorch/pytorch/blob/69f112d5867f785a3a090a0c6d6644ae047033ac/torch/_dynamo/symbolic_convert.py#L2060-L2071)
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -516,92 +507,91 @@ trace it and generate a graph generic on that variable.
 in size, you can mark it as dynamic by calling
-`torch._dynamo.mark_dynamic <https://github.com/pytorch/pytorch/blob/66a76516bfc341b2b55bb2056d2faa9c2de46d69/torch/_dynamo/decorators.py#L176>`__
-before calling ``torch.compile``. This will avoid the first compilation
+[torch.\_dynamo.mark_dynamic](https://github.com/pytorch/pytorch/blob/66a76516bfc341b2b55bb2056d2faa9c2de46d69/torch/_dynamo/decorators.py#L176)
```
```suggestion
[torch._dynamo.mark_dynamic](https://github.com/pytorch/pytorch/blob/66a76516bfc341b2b55bb2056d2faa9c2de46d69/torch/_dynamo/decorators.py#L176)
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -673,8 +660,8 @@ There is just one small issue with this. The statement “executing
-library? NumPy also has a Python API. Does ``torch.compile`` also
-understand NumPy? and Django? [5]_
+library? NumPy also has a Python API. Does `torch.compile` also
+understand NumPy? and Django? [^footnote-5]
```
```suggestion
understand NumPy? and Django? [^5]
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -686,15 +673,15 @@ The usual way machine learning tracers handle this issue is by informing
 code! To do this, Dynamo generates at trace time one graph with all the
 operations before the problematic code, and one with all the operations
-after. [6]_ Then, at runtime, it will delegate to CPython to execute the
+after. [^footnote-6] Then, at runtime, it will delegate to CPython to execute the
```
```suggestion
after. [^6] Then, at runtime, it will delegate to CPython to execute the
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -724,16 +711,17 @@ same arguments looks like this:
 
    1. It checks the guards from the first call against the new arguments
-      [`here <https://github.com/pytorch/pytorch/blob/e891a3bba9f05697d72776f6e89347231a141f03/torch/csrc/dynamo/eval_frame.c#L658>`__].
+      \[[here](https://github.com/pytorch/pytorch/blob/e891a3bba9f05697d72776f6e89347231a141f03/torch/csrc/dynamo/eval_frame.c#L658)\].
```
```suggestion
      [here](https://github.com/pytorch/pytorch/blob/e891a3bba9f05697d72776f6e89347231a141f03/torch/csrc/dynamo/eval_frame.c#L658).
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -839,30 +826,31 @@ breakpoints on them and looking at the call stack helps understanding
+Below are additional details and references for concepts mentioned in this document.
 
-.. [1] In the literature, this is called a Directed Acyclical Graph (DAG).
+[^footnote-1]: In the literature, this is called a Directed Acyclical Graph (DAG).
```
```suggestion
[^1]: In the literature, this is called a Directed Acyclical Graph (DAG).
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -839,30 +826,31 @@ breakpoints on them and looking at the call stack helps understanding
+[^footnote-1]: In the literature, this is called a Directed Acyclical Graph (DAG).
 
-.. [2] All this binding code lives in ``torch/csrc/dynamo/eval_frame.c``.
+[^footnote-2]: All this binding code lives in `torch/csrc/dynamo/eval_frame.c`.
```
```suggestion
[^2]: All this binding code lives in `torch/csrc/dynamo/eval_frame.c`.
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -839,30 +826,31 @@ breakpoints on them and looking at the call stack helps understanding
 
-.. [3] In CPython lingo, the set of all these objects are called `a
-   frame <https://github.com/python/cpython/blob/f26bfe4b25f7e5a4f68fcac26207b7175abad208/Include/internal/pycore_frame.h#L57-L71>`__.
+[^footnote-3]: In CPython lingo, the set of all these objects are called [a
```
```suggestion
[^3]: In CPython lingo, the set of all these objects are called [a
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -839,30 +826,31 @@ breakpoints on them and looking at the call stack helps understanding
 
-.. [4] There are also ``SymBool`` and ``SymFloat`` classes. The latter one
-   is not used all that much at the time of this writing.
+[^footnote-4]: There are also `SymBool` and `SymFloat` classes. The latter one
```
```suggestion
[^4]: There are also `SymBool` and `SymFloat` classes. The latter one
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -839,30 +826,31 @@ breakpoints on them and looking at the call stack helps understanding
-   and `the docs <https://pytorch.org/docs/main/torch.compiler_faq.html#does-numpy-work-with-torch-compile>`__.
-   Now, this is just possible because we reimplemented NumPy using
-   PyTorch. Good luck implementing Django in PyTorch though…
+[^footnote-5]: Interestingly enough, it does understand NumPy code! Have a look at
```
```suggestion
[^5]: Interestingly enough, it does understand NumPy code! Have a look at
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -724,16 +711,17 @@ same arguments looks like this:
+
    3. It gives CPython this new bytecode and asks it to run it
-      [`here <https://github.com/pytorch/pytorch/blob/e891a3bba9f05697d72776f6e89347231a141f03/torch/csrc/dynamo/eval_frame.c#L1006>`__]
+      \[[here](https://github.com/pytorch/pytorch/blob/e891a3bba9f05697d72776f6e89347231a141f03/torch/csrc/dynamo/eval_frame.c#L1006)\]
```
```suggestion
      \[[here](https://github.com/pytorch/pytorch/blob/e891a3bba9f05697d72776f6e89347231a141f03/torch/csrc/dynamo/eval_frame.c#L1006)\]
```

svekars commented on docs/source/torch.compiler_dynamo_deepdive.md:
```diff
@@ -29,89 +27,89 @@ executes the function and records a linear sequence of instructions
-and stores them sequentially. For example, it split ``z = (x - y) ** 2``
-into its two constituting operations, ``sub = l_x_ - l_y_`` and
-``z = sub ** 2``.
+of PyTorch operations. [^footnote-1] Dynamo records all the PyTorch operations
```
```suggestion
of PyTorch operations. [^1] Dynamo records all the PyTorch operations
```

svekars committed:
Apply suggestions from code review

svekars commented on docs/source/torch.compiler_custom_backends.md:
```diff
@@ -0,0 +1,279 @@
+Besides the `register_backend` decorator, if your backend is in another python package, you could also register your
+backend through entry points of python package, which provides a way for a package to register a plugin for another one.
+
+**Hint**\
```
a bit [broken](https://docs-preview.pytorch.org/pytorch/pytorch/155137/torch.compiler_custom_backends.html#registering-custom-backends) compared to [main](https://docs.pytorch.org/docs/main/torch.compiler_custom_backends.html#registering-custom-backends).
```suggestion
:::{hint}
```

svekars commented on docs/source/torch.compiler_custom_backends.md:
````diff
@@ -0,0 +1,279 @@
+    ]
+    ...
+)
+```
````
```suggestion
:::
```

svekars committed:
Update docs/source/torch.compiler_custom_backends.md

svekars committed:
Update docs/source/torch.compiler_custom_backends.md

svekars committed:
Update torch.compiler_custom_backends.md

svekars commented:
@pytorhcbot merge -f "Docs only change. Doc signal is green."

svekars commented:
@pytorchbot merge -f "Docs only change. Doc signal is green."
